[HttpFoundation] Fix the build on windows (with mbstring extension missing)

jakzal opened the pull request:
| Q | A |
| --- | --- |
| Branch | 2.3 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

An alternative approach to #16656 that works without the mbstring extension. If we tried to follow the current approach and make it work without mbstring it would make code quite more complex.

jakzal committed:
[HttpFoundation] Fix the build on windows (with mbstring extension missing)

jakzal commented:
Hm.. the issue is not fully fixed just yet. 

Before the build was failing with two errors:

```
Testing src/Symfony/Component/HttpFoundation
..............................E..
Fatal error: Call to undefined function Symfony\Component\HttpFoundation\mb_detect_encoding() in C:\projects\symfony\src\Symfony\Component\HttpFoundation\BinaryFileResponse.php on line 164
PHP Fatal error:  Call to undefined function Symfony\Component\HttpFoundation\mb_detect_encoding() in C:\projects\symfony\src\Symfony\Component\HttpFoundation\BinaryFileResponse.php on line 164
KO src/Symfony/Component/HttpFoundation
``` 
```
1) Symfony\Component\HttpFoundation\Tests\BinaryFileResponseTest::testConstructWithNonAsciiFilename
Symfony\Component\HttpFoundation\File\Exception\FileNotFoundException: The file "C:\projects\symfony\src\Symfony\Component\HttpFoundation\Tests/Fixtures/föö.html" does not exist
 
C:\projects\symfony\src\Symfony\Component\HttpFoundation\File\File.php:37
C:\projects\symfony\src\Symfony\Component\HttpFoundation\BinaryFileResponse.php:93
C:\projects\symfony\src\Symfony\Component\HttpFoundation\BinaryFileResponse.php:52
C:\projects\symfony\src\Symfony\Component\HttpFoundation\Tests\BinaryFileResponseTest.php:39
```

With this PR it only fails with the second one.


@xabbuh I'm not sure why it was earlier implemented with a loop through all the characters. Perhaps you could give some more insights?

WouterJ commented:
Any reason not to use https://github.com/symfony/polyfill-mbstring as it provides a polyfill for the mb functions used here?

jakzal commented:
@WouterJ that's an option indeed

jakzal committed:
[HttpFoundation] Fix the test on windows

A file with non-ascii characters in the name needs to be created from PHP
in order for it to be read properly on all platforms.

jakzal commented:
Tests fixed. Just bare in mind this is an alternative to #18030.

status: needs review

jakzal commented:
Actually, this will only work with unicode, while the existing solution supports all available encodings.
